compel/test/stack/spy.c

```diff
@@ -0,0 +1,368 @@
+
+	if (atexit(cleanup_saved_data))
+		err_and_ret("Can't register cleanup function with atexit\n");
+
+	stack = get_parasite_rstack_start(ctl);
+	if (save_data_near_stack(ctl, pid, stack, &saved_data, &saved_data_size))
+		err_and_ret("Can't save data above stack\n");
+
+	if (compel_start_daemon(ctl))
+		err_and_ret("Can't start daemon in victim\n");
+
+	/*
+	 * Now get the area with arguments and run two
+	 * commands one by one.
+	 */
+	arg = compel_parasite_args(ctl, int);
+
+	printf("Running cmd 1\n");
+	*arg = 137;
+	if (compel_rpc_call_sync(PARASITE_CMD_INC, ctl))
+		err_and_ret("Can't run parasite command 1\n");
+
+	printf("Running cmd 2\n");
+	*arg = 404;
+	if (compel_rpc_call_sync(PARASITE_CMD_DEC, ctl))
+		err_and_ret("Can't run parasite command 2\n");
+
+	if (check_saved_data(ctl, pid, stack, saved_data, saved_data_size))
+		err_and_ret("Saved data unexpectedly modified\n");
+
+	/*
+	 * Done. Cure and resume the task.
+	 */
+	printf("Curing\n");
+	if (compel_cure(ctl))
+		err_and_ret("Can't cure victim\n");
+
+	if (compel_resume_task(pid, state, state))
+		err_and_ret("Can't unseize task\n");
+
+	printf("Done\n");
+
+	return 0;
+}
+
+static inline int chk(int fd, int val)
+{
+	int v = 0;
+
+	if (read(fd, &v, sizeof(v)) != sizeof(v))
+		return 0;
+
+	printf("%d, want %d\n", v, val);
+	return v == val;
+}
+
+int main(int argc, char **argv)
+{
+	int p_in[2], p_out[2], p_err[2], pid, i, pass = 1;
+
+	/*
+	 * Prepare IO-s and fork the victim binary
+	 */
+	if (pipe(p_in) || pipe(p_out) || pipe(p_err)) {
+		perror("Can't make pipe");
+		return -1;
+	}
+
+	pid = vfork();
+	if (pid == 0) {
+		close(p_in[1]);
+		dup2(p_in[0], 0);
+		close(p_in[0]);
+		close(p_out[0]);
+		dup2(p_out[1], 1);
+		close(p_out[1]);
+		close(p_err[0]);
+		dup2(p_err[1], 2);
+		close(p_err[1]);
+		execl("./victim", "victim", NULL);
+		exit(1);
+	}
+
+	close(p_in[0]);
+	close(p_out[1]);
+	close(p_err[1]);
+
+	/*
+	 * Tell the little guy some numbers
+	 */
+	i = 1;
+	if (write(p_in[1], &i, sizeof(i)) != sizeof(i))
+		return 1;
+	i = 42;
+	if (write(p_in[1], &i, sizeof(i)) != sizeof(i))
+		return 1;
+
+	printf("Checking the victim alive\n");
+	pass = chk(p_out[0], 1);
+	pass = chk(p_out[0], 42);
+	if (!pass)
+		return 1;
+
+	/*
+	 * Now do the infection with parasite.c
+	 */
+
+	printf("Infecting the victim\n");
+	if (do_infection(pid))
+		return 1;
+
+	/*
+	 * Tell the victim some more stuff to check it's alive
+	 */
+	i = 1234;
+	if (write(p_in[1], &i, sizeof(i)) != sizeof(i))
+		return 1;
+	i = 4096;
+	if (write(p_in[1], &i, sizeof(i)) != sizeof(i))
+		return 1;
+
+	/*
+	 * Stop the victim and check the infection went well
+	 */
+	printf("Closing victim stdin\n");
+	close(p_in[1]);
+	printf("Waiting for victim to die\n");
+	wait(NULL);
+
+	printf("Checking the result\n");
+
+	/* These two came from parasite */
+	pass = chk(p_out[0], 138);
+	pass = chk(p_out[0], 403);
+
+	/* These two came from post-infect */
+	pass = chk(p_out[0], 1234);
+	pass = chk(p_out[0], 4096);
```

Comment:
Same here, previous chk()-s can both pass and fail and test would not mention this.

Reply:
Fixed in both tests.